File: Gold이상문제_정리/ACM_craft_1005.java
package Gold이상문제_정리;

import java.util.*;
import java.io.*;

public class ACM_craft_1005 {
	static int T, N, K;
	static int[] D;
	static int [] countDegree;
	static ArrayList<ArrayList<Integer>> adj;
	static Queue<Integer> que = new LinkedList<>();
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		StringBuilder str = new StringBuilder();

		T = Integer.parseInt(st.nextToken());
		while (T-- > 0) {
			int time = play();
			str.append(time).append("\n");
		}
		System.out.println(str);
	}
	private static int play() throws IOException {
		que.clear();
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());

		//작업 시간
		st = new StringTokenizer(br.readLine());
		D = new int[N+1];
		for (int n = 1; n <= N; n++)
			D[n] = Integer.parseInt(st.nextToken());

		// 연관된 작업
		countDegree = new int[N + 1];
		adj = new ArrayList<>();
		for (int i = 0; i <= N; i++)
			adj.add(new ArrayList<>());

		// 작업 연결
		for (int k = 0; k < K; k++) {
			st = new StringTokenizer(br.readLine());
			int first = Integer.parseInt(st.nextToken());
			int after = Integer.parseInt(st.nextToken());
			adj.get(first).add(after); // 1 - 2, 3, 2 4, 34
			countDegree[after]++;// 2 ++, 3++, 4++, 4++
			// 먼저 해야될 조건들을 뒤에 달아준다
		}

		// 마지막 작업
		st = new StringTokenizer(br.readLine());
		int W = Integer.parseInt(st.nextToken());

		for (int i = 1; i <= N; i++) {
			if (countDegree[i] == 0) {
				que.add(i);
			}
		}

		int[] timeArr = new int[N+1];
		Arrays.fill(timeArr, 0);
		while (countDegree[W] > 0 && !que.isEmpty())
		{
			int job = que.poll(); // 먼저 할거
			for (int nextJob : adj.get(job)) { // job에 연결된 다음 작업 실시
				timeArr[nextJob] = Math.max(timeArr[nextJob], timeArr[job] + D[job]);
				countDegree[nextJob]--;
				if (countDegree[nextJob] == 0)
					que.add(nextJob);
			}
		}
		timeArr[W] += D[W];
		return timeArr[W];
	}
}

File: 코틀린_코테스터디/week4/ACMcraft_1005.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max


private var T = 0
private var N = 0
private var K = 0
private lateinit var D: IntArray
private lateinit var countDegree: IntArray
private var adj: ArrayList<ArrayList<Int>> = ArrayList()
private var que: Queue<Int> = LinkedList()
private var br = BufferedReader(InputStreamReader(System.`in`))
private var st: StringTokenizer? = null

fun main() {
    st = StringTokenizer(br.readLine())
    val str = StringBuilder()
    T = st!!.nextToken().toInt()

    while (T-- > 0) {
        val time = play()
        str.append(time).append("\n")
    }

    println(str)
}

private fun play(): Int {
    que.clear()
    st = StringTokenizer(br.readLine())
    N = st!!.nextToken().toInt()
    K = st!!.nextToken().toInt()

    //작업 시간
    st = StringTokenizer(br.readLine())
    D = IntArray(N + 1)
    for (n in 1..N) {
        D[n] = st!!.nextToken().toInt()
    }

    // 연관된 작업
    countDegree = IntArray(N + 1)
    adj = ArrayList()
    for (i in 0..N) {
        adj.add(ArrayList())
    }

    // 작업 연결
    for (k in 0 until K) {
        st = StringTokenizer(br.readLine())
        val first = st!!.nextToken().toInt()
        val after = st!!.nextToken().toInt()
        adj[first].add(after)
        countDegree[after]++
    }

    // 마지막 작업
    st = StringTokenizer(br.readLine())
    val w = st!!.nextToken().toInt()
    for (i in 1..N) {
        if (countDegree[i] == 0) {
            que.add(i)
        }
    }

    val timeArr = IntArray(N + 1)
    Arrays.fill(timeArr, 0)

    while (countDegree[w] > 0 && !que.isEmpty()) {
        val job = que.poll() // 먼저 할거
        for (nextJob in adj[job]) { // job에 연결된 다음 작업 실시
            timeArr[nextJob] = max(timeArr[nextJob], timeArr[job] + D[job])
            countDegree[nextJob]--
            if (countDegree[nextJob] == 0) que.add(nextJob)
        }
    }

    timeArr[w] += D[w]
    return timeArr[w]
}
